### ISSUE TITLE
Store interval frequency of data points in Datetime column

### ISSUE THREAD

**Issue by tamargrey** [alteryx/woodwork#1155]
In certain instances, it is useful to know the frequency of observations in a Datetime column (1 min, 1 day, 5 days) where the interval of time between data points in the column is consistent. These frequencies may be known by users or inferred from the data. 

As a user, it could be useful for down-the-line operations that will try to infer that frequency or otherwise require it be passed in as an argument (for example: pandas' [to_period](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dt.to_period.html)) if a Woodwork DataFrame had knowlege of a Datetime column's frequency. 

We should consider storing a column's frequency as an [offset string](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases) or offset object.

#### Code Example

Storing the frequency in the `Datetime` logical type:

```python
df.ww.init(logical_types={'dates': Datetime(frequency='1D')})
assert df.ww.logical_types['dates'].frequency == '1D'
```

This may be a relevant concept only for time index, so it could also be a value that's inferred via `pd.infer_freq` just for the time index column:

```python
df.ww.init(time_index='dates')})
assert df.ww.logical_types['dates'].frequency == '1D'
```

Additionally, since Woodwork allows numeric time indexes, is this a concept that can or should exist for a numeric time index?

Questions to answer:
- What does it look like if a column has no discernable frequency and one is assigned or attempted to be inferred? 
- What if a few columns are not compliant with the inferred or assigned frequency?
- Does it matter if the data is sorted or not?
- Can users assign a frequency after woodwork initialization?

**Comment by gsheni** [comment]
Part of this issue will be determining if we can infer this time interval or if we should punt on that part.

**Comment by dsherry** [comment]
Yep!

I have a couple more questions for consideration.

Q: What's the initial set of time series units this detection will support?

My advice is that ['YEAR', 'MONTH', 'WEEK', 'DAY', 'HOUR', 'SECOND', 'MILLISECOND'] is a good starting point, tagging @freddyaboulton @ParthivNaresh for comment. Ideally woodwork can also define an enum to denote the unit, as well as a string representation of that enum. 

Q: How important is it to detect both the fundamental unit each row is spaced by and the number of units? Ex: if each row of data is spaced by exactly two days, is it important that we know the unit is "DAY" and the number of units is 2, or is it enough to simply know that the unit is "DAY"?

My advice is to start with only detecting the lowest possible granularity required to represent the data in question. One idea on an algo to compute this:
* Compute the time delta of each row vs the preceding row
* Take the min of the deltas
* Round that down to the nearest supported time unit. E.g. if that's two days, the nearest supported would be "DAY"
* Return that unit

**Comment by freddyaboulton** [comment]
I think it would be best if the frequency stored in the time index corresponded to the time delta between rows and not the fundamental unit as proposed by @dsherry. So in code:

```python
df = pd.DataFrame({"date": pd.Series(pd.date_range("2021-01-01", periods=5, freq="2W"))})
df.ww.init(time_index="date")
assert df.ww.logical_types['date'].freq == '2W'
```

I think there are two use-cases where it would be important to have this behavior:

1. Data Validation: I have a done some feature engineering and trained a machine learning model on a training set. I need to know that the test set has the same time delta between rows as my training set otherwise my feature engineering will be off. I might also need to verify the test set is separated from the training set by a given amount of time. If we just store the "fundamental unit" this would not be possible, because the training set could have a delta of 2 days, and the test set could have a delta of one day, but both would have the same woodwork `freq`
2. Communicating intended feature engineering behavior to users: Right now, EvalML and the featuretools RollingWindow transform primitives work off of integer-offsets. The integer offset corresponds to a multiple of the time delta between rows. If the data has a frequency of 2 days, but we show it to users as a single day, when they parametrize their RollingMean/Lag transforms, and use a window size of `5`, they might think that window size corresponds to 5 days. In reality it'd be 10 days.

**Comment by tamargrey** [comment]
@freddyaboulton Are we okay with just inferring and storing the frequency for the time index column? Would it be desirable, if not a requirement right now, to allow frequencies to be set for non-time index datetime columns?

**Comment by freddyaboulton** [comment]
I think the use cases we're primarily interested would only care/use the frequency of the time index column so I think it's ok if we just store it for the time index column!

**Comment by ParthivNaresh** [comment]
+1 regarding Freddy's comments

What does it look like if a column has no discernable frequency and one is assigned or attempted to be inferred?

We could force the time index column to follow the interval we've inferred i.e. starting from the first time value, we project forward the expected time values based on the interval we've inferred. Any time values not matching will be replaced by what we projected. This is an abrasive option and might be tricky with out of sample data. We'd have to make sure it follows the same interval as what was inferred from the training data, possibly taking into account a separation of `gap`.

What if a few columns are not compliant with the inferred or assigned frequency?

Not sure what this means compared to the previous question.

Does it matter if the data is sorted or not?

We require the data to be monotonically increasing and have equal intervals as detailed in `DateTimeFormatDataCheck`. But this is more relevant for the `date_index` column. If we're looking at any datetime column as @gsheni mentioned, then sorting it would be necessary behind the scenes to find the proper interval, but the final feature would be returned how it was originally sorted.

Can users assign a frequency after woodwork initialization?

I would say no for now, at least in the scope of this issue. If we want to expand this in the future to allow users to change the frequency on the fly, then we'll have to change the time values alongside to match the frequency they've passed. I'm not sure how allowing users to set their own frequency would work with datetime features that are unsorted.

**Comment by tamargrey** [comment]
Thanks @ParthivNaresh! 

At this point, it seems like the simplest thing we can do is not perform frequency inference at Woodwork init or store inferred frequencies within the Woodwork typing information.

Instead, we can start with a Woodwork method that will call `infer_freq` on the dataframe's datetime columns (or on a specific column if the time index is known). Note that this means we'd be inferring the "frequency stored in the time index corresponded to the time delta between rows and not the fundamental unit" (@freddyaboulton @dsherry). 

Not storing the frequency solves a few of the questions I've been kicking around about validating "assigned" frequencies or keeping them updated if the data is changed. A woodwork user would be in charge of storing these if they need to reference back to them. If, at some point in the future, we want to store the frequency information, we can handle the validation then. 

Does that seem like a fair approach for now?

**Comment by freddyaboulton** [comment]
@tamargrey Thanks for the explanation! Is there a mechanism for users to store this information in the schema? Like in the metadata or something? With the assumption the user would be in charge of keeping this up to date. Is it possible to add this information to the typing information in the future? 

I'm asking cause I envision we'll have to access this information in the not too distant future in multiple parts of the code and it might make sense to store this along with the data. For the kinds of problems I'm interested in at least, the frequency should not change once it's inferred/set. 

Not storing along with the data is not a blocker but just wondering.

**Comment by tamargrey** [comment]
@freddyaboulton 

> Is there a mechanism for users to store this information in the schema? Like in the metadata or something? 

Frequency could absolutely be stored in the column metadata! But that would definitely be something for a user to decide since we don't want Woodwork to set any of its own metadata.

If Woodwork was going to be in charge of storing its column frequencies, we'd probably store it on the `Datetime` logical type. But this would require us to validate `frequency` in much the same way we validate `datetime_format`, since a user could take a Datetime logical type object and try and use it for a different column.

**Comment by freddyaboulton** [comment]
Thanks @tamargrey !
